backend/lib/endpoints/organizations.js

```diff
@@ -0,0 +1,110 @@
+const httpErrors = require("http-errors");
+
+const {
+  createOrganizationSchema,
+  deleteOrganizationSchema,
+  getOrganizationSchema,
+  getOrganizationsSchema,
+  updateOrganizationSchema,
+} = require("./schema/organizations");
+
+/*
+ * /api/organizations
+ */
+async function routes(app) {
+  const Organization = app.mongo.model("Organization");
+
+  app.delete(
+    "/:organizationId",
+    {
+      preValidation: [app.authenticate],
+      schema: deleteOrganizationSchema,
+    },
+    async (req) => {
+      // TODO: make sure user can only delete organizations they own
+      const result = await Organization.findByIdAndRemove(
+        req.params.organizationId,
+      );
+      if (result === null) {
+        return new httpErrors.NotFound();
+      }
+      return result;
+    },
+  );
+
+  app.get(
+    "/",
+    {
+      preValidation: [app.authenticate],
+      schema: getOrganizationsSchema,
+    },
+    async (req) => {
+      const { ownerId } = req.params;
+      const filter = ownerId ? { ownerId } : {};
+      const sortedOrganizations = await Organization.find(filter).sort({
+        name: 1,
+      });
+      return sortedOrganizations;
+    },
+  );
+
+  app.get(
+    "/:organizationId",
+    {
+      preValidation: [app.authenticate],
+      schema: getOrganizationSchema,
+    },
+    async (req) => {
+      const result = await Organization.findById(req.params.organizationId);
+      if (result === null) {
+        return new httpErrors.NotFound();
+      }
+      return result;
+    },
+  );
+
+  app.patch(
+    "/:organizationId",
+    {
+      preValidation: [app.authenticate],
+      schema: updateOrganizationSchema,
+    },
+    async (req) => {
+      // TODO: make sure user can only update organizations they own
+      const organization = await Organization.findById(
+        req.params.organizationId,
+      );
+      if (organization === null) {
+        return new httpErrors.NotFound();
+      }
+      const updateProperties = updateOrganizationSchema.body.properties;
+      const errors = [];
+      Object.keys(req.body).forEach((key) => {
+        if (!Object.prototype.hasOwnProperty.call(updateProperties, key)) {
```

Comment:
Since there's now `Object.prototype.hasOwnProperty()` you probably don't have to use `. call` but you can use this directly on `updateProperties.hasOwnProperty(key)` 

Before it was more like a generic `Object.hasOwnProperty(key)` but now that it's on the prototype it can be used on any instance of an object, also on a literal.

Also... hmm we are kind of doing the same work that the JSON schema has already done when the request is validated because we use the same schema anyway.

Reply:
> Since there's now Object.prototype.hasOwnProperty() you probably don't have to use .call() but you can use this directly on updateProperties.hasOwnProperty(key)

I actually had that initially (`updateProperties.hasOwnProperty(key)` but the linter didn't like it:
```
83:31  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins
```
`Object.prototype.hasOwnProperty.call` was the recommended by [ESLint](https://eslint.org/docs/rules/no-prototype-builtins). ¯\_(ツ)_/¯ 


> Also... hmm we are kind of doing the same work that the JSON schema validation has already done when the request is validated because we use the same schema anyway.

Based on my testing, when I had it the way it was before, it just ignored extra properties not defined on the schema, and it did not return a 400 bad request error. Is this what we want? I also noticed that sending a post request with extra properties also ignored them.

Reply:
Hmmm... 
Besides, maybe we should just make sure every schema has the "additionaProperties: false" setting, I forgot about that, in JSON schema you probably have to set that explicitly.
Unless we use something like this https://github.com/greguz/fastify-no-additional-properties which enables it by default for all schema validation in fastify.

Reply:
I think I tried adding that the FluentSchema but it didn't work. But I'll try again tonight to confirm.

Reply:
Okay, finally figured this out. So by default, Fastify removes additional properties by setting the `removeAdditional` ajv config option to `true` (see https://www.fastify.io/docs/latest/Validation-and-Serialization/). So by the time it reaches the route handler, the property is no longer there.

If I override the default configuration to set the `removeAdditional` config option to `false`, and also add `additionalProperties(false)` to the Fluent Schema, I now get the 400 bad request error:

```
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": "body should NOT have additional properties"
}
```

Reply:
@robinv85 check out my latest commits. I reverted my commit where I was checking for additional properties in the route handler, added an overridden ajv config, and also set `additionalProperties` to false on the schemas.

I did try that library you mentioned where it would automatically set `additionalProperties` to false for all schemas, but it didn't work at all. Maybe we can try to roll our own, but I'll leave that for another PR.

Reply:
Hmm ok so I guess that's more explicit to return a bad request error instead of removing the additional properties. 
Only disadvantage is that it's a risk to forget this in the schema. For me this is ok but would be nice if we can set it as a default without risking to forget it in the schema's.
Can you also add this in the readme?

Reply:
Will add to the readme tonight after work. 

Yes, I think it's easier to debug in case a property was misspelled. Better to fail fast than silently.

Reply:
I updated the readme with this caveat, but we should also look into automatically adding this to all our schemas. I did try the library you mentioned but unfortunately I was not able to get it working. We can perhaps address this in another ticket. I will add a ticket for it.